### PR TITLE
fix consumer subscribe bug

### PR DESCRIPTION
**Pull request by rkatsyuryna**

### CODE REVIEW

File: core/src/main/java/io/github/tcdl/collector/Collector.java [hunk 1]
```diff
@@ -49,23 +46,30 @@
     private TimeoutManager timeoutManager;
 
     private Clock clock;
-
+    private MsbContextImpl msbContext;
```
Comment (anstr) [inline core/src/main/java/io/github/tcdl/collector/Collector.java]: I'm not sure that put msbContext inside Collector is good idea,as I understand MsbContext holds everything for creating Requester,ResponderServer etc..So with MsbContext  we create  Requester and it holds Collector with MsbContext as a field,very complex)
Also "private MsbContextImpl msbContext;"=>"private MsbContext msbContext;"

File: core/src/main/java/io/github/tcdl/collector/CollectorManagerFactory.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package io.github.tcdl.collector;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import io.github.tcdl.ChannelManager;
+
+public class CollectorManagerFactory {
```
Comment (anstr) [inline core/src/main/java/io/github/tcdl/collector/CollectorManagerFactory.java:8]: Not sure that CollectorManagerFactory  is right name.

File: core/src/main/java/io/github/tcdl/collector/Collector.java (hunk 1 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/collector/Collector.java]: removed

File: core/src/main/java/io/github/tcdl/collector/CollectorManagerFactory.java (hunk 2 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/collector/CollectorManagerFactory.java:8]: suggest better ?

File: core/src/main/java/io/github/tcdl/ChannelManager.java [hunk 3]
```diff
@@ -53,10 +53,22 @@ public Producer findOrCreateProducer(final String topic) {
         return producer;
     }
 
-    public Consumer findOrCreateConsumer(final String topic) {
+    /**
+     * Start consuming messages on specified topic with handler .
+     * Calls to subscribe() and unsubscribe() have to be properly synchronized by client code not to lose messages.
```
Comment (vso-tc) [inline core/src/main/java/io/github/tcdl/ChannelManager.java]: I don't completely understand why do we need synchronization between subscribe and unsubscribe.

File: core/src/main/java/io/github/tcdl/Consumer.java [hunk 4]
```diff
@@ -32,35 +32,31 @@
     private ChannelMonitorAgent channelMonitorAgent;
     private Clock clock;
 
+    //handler that invoked when message being received.
```
Comment (vso-tc) [inline core/src/main/java/io/github/tcdl/Consumer.java]: Let's use javadoc syntax here.

File: core/src/main/java/io/github/tcdl/adapters/AdapterFactory.java [hunk 5]
```diff
@@ -12,7 +12,7 @@
     
     /**
      * Initialize AdapterFactory. The method should be called only once from AdapterFactoryLoader.
-     * @param msbConfig - MsbConfigurations object 
+     * @param msbConfig - MsbConfig object
```
Comment (vso-tc) [inline core/src/main/java/io/github/tcdl/adapters/AdapterFactory.java]: @link would be nice here

File: core/src/main/java/io/github/tcdl/collector/CollectorManagerFactory.java (hunk 2 above)
Comment (vso-tc) [inline core/src/main/java/io/github/tcdl/collector/CollectorManagerFactory.java:8]: Suggesting revolution:
CollectorManager -> CollectorTopicManager
CollectorManagerFactory -> CollectorManager

File: core/src/test/java/io/github/tcdl/ChannelManagerConcurrentTest.java [hunk 6]
```diff
@@ -57,6 +55,8 @@ public void testConsumerCachedMultithreadInteraction() throws Exception {
     }
 
     @Test
+    @Deprecated
```
Comment (vso-tc) [inline core/src/test/java/io/github/tcdl/ChannelManagerConcurrentTest.java]: It's better to leave a comment why do we ignore or deprecate some things

File: core/src/main/java/io/github/tcdl/collector/Collector.java [hunk 7]
```diff
@@ -49,23 +46,25 @@
...
-    private CollectorManager collectorManager;
+    private  CollectorManager collectorManager;
+
+    public Collector(String topic, Message requestMessage, RequestOptions requestOptions, MsbContextImpl msbContext, EventHandlers eventHandlers) {
+        this.requestMessage = requestMessage;
 
-    public Collector(RequestOptions requestOptions, MsbContextImpl msbContext, EventHandlers eventHandlers) {
-        this.channelManager = msbContext.getChannelManager();
         this.clock = msbContext.getClock();
+        this.collectorManager = msbContext.getCollectorManagerFactory().findOrCreateCollectorManager(topic);
```
Comment (vso-tc) [inline core/src/main/java/io/github/tcdl/collector/Collector.java:64]: Couldn't manager be injected here? This way Collector would not create its CollectorManager

File: core/src/main/java/io/github/tcdl/collector/CollectorManagerFactory.java (hunk 2 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/collector/CollectorManagerFactory.java:8]: not sounds nicer(

File: core/src/main/java/io/github/tcdl/Consumer.java (hunk 4 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/Consumer.java]: ok

File: core/src/main/java/io/github/tcdl/adapters/AdapterFactory.java (hunk 5 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/adapters/AdapterFactory.java]: ok

File: core/src/main/java/io/github/tcdl/collector/Collector.java (hunk 7 above)
Comment (rkatsyuryna) [inline core/src/main/java/io/github/tcdl/collector/Collector.java:64]: not make sense for me, Collector constructor is already having a lot parameters.

File: core/src/test/java/io/github/tcdl/ChannelManagerConcurrentTest.java (hunk 6 above)
Comment (rkatsyuryna) [inline core/src/test/java/io/github/tcdl/ChannelManagerConcurrentTest.java]: will remove this test. it's not more useful

**Pull request merged**